### DISCUSSION TITLE
Intended behavior of explicit IDisposable.Dispose implementations in ref structs.

### DISCUSSION THREAD

**Discussion by aloraman** [dotnet/csharplang#8405] [General]
**Disclaimer:** Originally posted at [runtime repo](https://github.com/dotnet/runtime/issues/107140), as I was not sure if it was intended language design, a compiler bug, or a general compatibility-driven decision.

One of C# 13 features announced in [What's new in C# 13](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13#ref-struct-interfaces) is to enable ref structs to implement interfaces. It allows ref structs to implement IDisposable interface, moreover it allows them to implement IDisposable explicitly. Currently, when implemented both implicitly and explicitly, ref structs behave differently to other types, e.g. consider the following example:

```csharp
using System;

TestClass(new FooClass()); // Explicit
TestStruct(new FooStruct()); // Explicit
TestRefStruct(new FooRefStruct()); // Implicit!

void TestClass(FooClass target) {using(target){}}
void TestStruct(FooStruct target) {using(target){}}
void TestRefStruct(FooRefStruct target) {using(target){}}


public class FooClass: IDisposable
{
  public void Dispose() => Console.WriteLine("Implicit");
  void IDisposable.Dispose() => Console.WriteLine("Explicit");
}

public struct FooStruct: IDisposable
{
  public void Dispose() => Console.WriteLine("Implicit");
  void IDisposable.Dispose() => Console.WriteLine("Explicit");
}

public ref struct FooRefStruct: IDisposable
{
  public void Dispose() => Console.WriteLine("Implicit");
  void IDisposable.Dispose() => Console.WriteLine("Explicit");
}
```
[Run at Sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQCoCmAzggMIA2EddAFDDQO4AEAMQD2Qlm04BKCQG4+Aejl8AogA8ADsygBjKAmr0EAZQRotCLr0Ejjp81NkLl6zTr3laDAEo0AZjdRmFvzCQt5+JgF2MvKKAJJgGtq6AIQUFPgALHweTKzsHCFi7HwIEMgA5jQIEnwA3vgADBylFVUStQC+HelZOf6BIf0IJWWV1XWNzaNtnd3kmdkGYUMFIssRZiOt4/VNLWPtXWmU8wDMfERWonl0IHyxACJQdGpCdBDAzDQUtRR8F+cFk8Xm8aBwagBeAB8F2IAE4OAAieKJVyImR/C5ZR7PV7vT40AB0wLxYMhMPw8KRqlRunR0goc3S5wYtiuQzuOJB+K+P0x+EBWRJoPBfGhsIRyISLjpGPI/wWXLxHy+xNxIvJEupziSCHpjOOAr4yF8fFZkSu61snOFPO+5F+8oBWL4trJYopVKltL1coV2NtKqJbtF4spkppMt9DPIHSAA)

That is, ref structs prefer implicit* Dispose implementation, while every other type prefers explicit Dispose implementation, conforming to ECMA-334 spec. 

*_Colloquially called implicit, in other words - not explicit. For ref structs it means choosing a specific method by matching  `public void Dispose()` signature, i.e. duck typing._

Is this intended design or a bug? If former, would it be possible to reconsider this? Now, it creates a weird behavior, the only purpose of which is to be a nasty puzzler for a programmer interview. If changed, it would be a breaking change, but a very minor one. 
To observe such a breaking change, you'll need to explicitly add an IDisposable implementation to an existing ref struct with implicit Dispose method - which is somewhat of a breaking change in itself. Now, if such a struct is used in a Dispose statement - different overload will be called, so it is a source level breaking change, but it's no different from adding explicit Dispose implementation to any regular disposable type anyway. As far as I understand, this is not a binary breaking change - existing binaries that utilize ref struct Disposal have a direct call to implicit Dispose method on IL level anyway - so they will continue to call implicit implementations.

**NB**
As far as I understand how consistent the behavior is with `allow ref struct` feature (I don't have NET 9.0 version locally with this feature enabled), in generic code an explicit implementation will be used (even though JIT could've replaced `callvirt IDisposable.Dispose` to `call Dispose` due to per-concrete type reification for structs):

```csharp
TestGeneric(new FooClass()); // Explicit
TestGeneric(new FooStruct()); // Explicit
TestGeneric(new FooRefStruct()); // Explicit (!)

void TestGeneric<T>(T target) where T:IDisposable, allow ref struct => {using(target){}}
```
Which makes non-generic case even more inconsistent.

**Comment by HaloFour** [comment]
Before interfaces were supported on `ref struct`s the C# language allowed you to use `using` with those types as long as they had an accessible `Dispose` method on them.  The compiler still prefers that pattern for the sake of backward compatibility and explicitly emits a call to that `Dispose` method without even checking to see if the type implements `IDisposable`.  Whether or not the compiler should continue to do that is probably a conversation worth having.

```cs
using System;

var target = new FooRefStruct();
using (target) { };

public ref struct FooRefStruct
{
  public void Dispose() => Console.WriteLine("Implicit");
}
```

**Reply by CyrusNajmabadi** [reply]
Personally, if you have a ref struct with that would behave differently, then I think that's a bad thing.  So I would just stick with our current behavior and not change it. In practice it should not matter. And if it does, meh :-)

**Reply by 333fred** [reply]
Agreed with Cyrus. The behavior is a fallout of the rules exactly as written; while we _could_ change it, it really doesn't seem necessary to me. The only reason that `struct` and `class` don't behave the same way is because we're too afraid of making the breaking change to have them work this way. When we did `using` for `ref struct`s, we actually tried to extend pattern-based `using` to all types, but it caused breaking changes that we didn't want to take. `ref struct`s actually work the way we'd want any type to work in an ideal world.

**Reply by aloraman** [reply]
@CyrusNajmabadi Still, the preference of choosing `IDisposable.Dispose` (without boxing) made it into ECMA-334. I'd prefer ref structs to behave the same way as the other types - for the sake of consistency. Fewer language quirks, fewer brainteaser questions for an interview. Also, if ref struct could have *interface implementation* feature in the same language release as *disposal* feature - there would be no need for Dispose-like method matching in compiler, so I'd argue for changing current situation to get rid of *obsolete intermediary step*.

**Reply by CyrusNajmabadi** [reply]
> fewer brainteaser questions for an interview.

Honestly, I could not care less.  :-)

> so I'd argue for changing current situation to get rid of obsolete intermediary step.

I'd be fine with a larger, holistic, improvement to the whole disposal area. However, it would only be through that that I would change this here.  I would never touch this up front, thinking that we might do the other stuff later.

**Reply by jnm2** [reply]
> fewer brainteaser questions for an interview.

Potentially serving as a way to filter out interview***ers*** :D

**Comment by colejohnson66** [comment]
~~I'd argue that explicit interface implementations on ref structs don't make sense and should be prohibited. 
After all, you and the compiler can't reach them (they require boxing).~~

**Reply by jnm2** [reply]
They can be reached without boxing (using the .constrained IL prefix for the callvirt) via:
```cs
void M<T>(T disposable) where T : IDisposable, allows ref struct
{
    disposable.Dispose();
}
```

**Reply by aloraman** [reply]
Also, disposal of other types does not involve boxing, even though the `IDisposable.Dispose()` is prefered - as specified in ECMA-334 (once for `using` block, and once for `foreach` block again)

**Reply by jnm2** [reply]
[Oh I see, that is also a constrained call!](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIwAYACAChgFMB3MgDQoEpmyBvMgX0MICdqAMzIBnBHwCuAYwSMyIMgEkAIlBEAHAPYiIwFNULtCZE2QwAWJao3bd+gHTWtI6i2OmjBU97M4AnBQARE7a1IHMANzuJjwEXEA)

![image](https://github.com/user-attachments/assets/d31319cf-6a8d-47bd-b7e3-d92a2fe47e28)
